File: tools/zoom.py
#!/usr/bin/env python3
"""zoom.py — pick a fixed-grid chunk of a screenshot to look at closer.

The screen is divided into a 4-column x 4-row grid (16 chunks total).
Each chunk is 80x50 pixels for a 320x200 source. You name a chunk by
(col, row) where col is 0..3 (left to right) and row is 0..3 (top to
bottom). No rectangle picking, no upscaling — feeding a smaller crop
directly already gives more output tokens per source pixel.

In some programs the rows usually carry:
  row 0: sky / mountain tops
  row 1: tall scenery (building bodies, tree tops)
  row 2: walking band (player and door slots live here)
  row 3: HUD strip + foreground plants

Column 0..3 is left to right of the visible play area.

Usage:
  zoom.py SRC COL ROW [--out PATH]
"""
from __future__ import annotations

import argparse
import sys
from pathlib import Path

try:
    from PIL import Image
except ImportError:
    sys.exit("zoom: needs Pillow — `pip install Pillow`")

GRID_COLS = 4
GRID_ROWS = 4


def main() -> None:
    parser = argparse.ArgumentParser(
        description=__doc__,
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    parser.add_argument("src", help="source PNG to crop")
    parser.add_argument(
        "col", type=int,
        help=f"grid column 0..{GRID_COLS - 1} (left to right)")
    parser.add_argument(
        "row", type=int,
        help=f"grid row 0..{GRID_ROWS - 1} (top to bottom)")
    parser.add_argument(
        "--out", default=None,
        help="output path (default: <src>_zoom_<col>_<row><ext>)")
    args = parser.parse_args()

    if not (0 <= args.col < GRID_COLS) or not (0 <= args.row < GRID_ROWS):
        sys.exit(f"zoom: col must be 0..{GRID_COLS - 1}, "
                 f"row must be 0..{GRID_ROWS - 1}")

    src_path = Path(args.src)
    if not src_path.exists():
        sys.exit(f"zoom: source not found: {src_path}")
    img = Image.open(src_path).convert("RGB")
    sx, sy = img.size

    cw, ch = sx // GRID_COLS, sy // GRID_ROWS
    x_lo = args.col * cw
    y_lo = args.row * ch
    # Extend the right/bottom-most chunk to absorb the integer-division remainder.
    x_hi = sx if args.col == GRID_COLS - 1 else x_lo + cw
    y_hi = sy if args.row == GRID_ROWS - 1 else y_lo + ch

    crop = img.crop((x_lo, y_lo, x_hi, y_hi))
    out = Path(args.out) if args.out else src_path.with_name(
        f"{src_path.stem}_zoom_{args.col}_{args.row}{src_path.suffix}"
    )
    out.parent.mkdir(parents=True, exist_ok=True)
    crop.save(out)
    print(out)


if __name__ == "__main__":
    main()

File: saisei/src/zoom.rs
//! `saisei zoom` — crop a fixed-grid chunk of a screenshot to look closer.
//! (the `image` crate). The screen is a
//! 4-column x 4-row grid (16 chunks). You name a chunk by (col, row), col 0..3
//! left->right and row 0..3 top->bottom; the right/bottom-most chunk absorbs the
//! integer-division remainder. No upscaling — a smaller crop already yields more
//! output tokens per source pixel.

use std::path::{Path, PathBuf};
use std::process::exit;

const GRID_COLS: u32 = 4;
const GRID_ROWS: u32 = 4;

fn die(msg: &str) -> ! {
    eprintln!("{msg}");
    exit(1)
}

fn parse_int(s: &str, what: &str) -> i64 {
    s.parse::<i64>()
        .unwrap_or_else(|_| die(&format!("zoom: {what} must be an integer: {s}")))
}

pub fn main(_root: &Path, args: &[String]) -> ! {
    let mut positional: Vec<String> = Vec::new();
    let mut out: Option<String> = None;
    let mut it = args.iter();
    while let Some(a) = it.next() {
        match a.as_str() {
            "--out" => {
                out = Some(
                    it.next()
                        .cloned()
                        .unwrap_or_else(|| die("zoom: --out needs a path")),
                )
            }
            other if other.starts_with("--") => {
                die(&format!("zoom: unrecognized argument: {other}"))
            }
            other => positional.push(other.to_string()),
        }
    }
    if positional.len() < 3 {
        die("usage: zoom SRC COL ROW [--out PATH]");
    }
    let src = positional[0].clone();
    let col = parse_int(&positional[1], "col");
    let row = parse_int(&positional[2], "row");

    if !(0..GRID_COLS as i64).contains(&col) || !(0..GRID_ROWS as i64).contains(&row) {
        die(&format!(
            "zoom: col must be 0..{}, row must be 0..{}",
            GRID_COLS - 1,
            GRID_ROWS - 1
        ));
    }
    let col = col as u32;
    let row = row as u32;

    let src_path = PathBuf::from(&src);
    if !src_path.exists() {
        die(&format!("zoom: source not found: {}", src_path.display()));
    }
    let img = image::open(&src_path)
        .unwrap_or_else(|e| die(&format!("zoom: cannot open {}: {e}", src_path.display())))
        .to_rgb8();
    let (sx, sy) = img.dimensions();

    let cw = sx / GRID_COLS;
    let ch = sy / GRID_ROWS;
    let x_lo = col * cw;
    let y_lo = row * ch;
    // Extend the right/bottom-most chunk to absorb the integer-division remainder.
    let x_hi = if col == GRID_COLS - 1 { sx } else { x_lo + cw };
    let y_hi = if row == GRID_ROWS - 1 { sy } else { y_lo + ch };

    let crop = image::imageops::crop_imm(&img, x_lo, y_lo, x_hi - x_lo, y_hi - y_lo).to_image();

    let out_path = match out {
        Some(p) => PathBuf::from(p),
        None => {
            let stem = src_path
                .file_stem()
                .map(|s| s.to_string_lossy().into_owned())
                .unwrap_or_default();
            let suffix = src_path
                .extension()
                .map(|s| format!(".{}", s.to_string_lossy()))
                .unwrap_or_default();
            let name = format!("{stem}_zoom_{col}_{row}{suffix}");
            match src_path.parent() {
                Some(p) if !p.as_os_str().is_empty() => p.join(name),
                _ => PathBuf::from(name),
            }
        }
    };

    if let Some(parent) = out_path.parent() {
        if !parent.as_os_str().is_empty() {
            std::fs::create_dir_all(parent).ok();
        }
    }
    crop.save(&out_path)
        .unwrap_or_else(|e| die(&format!("zoom: cannot save {}: {e}", out_path.display())));
    println!("{}", out_path.display());
    exit(0);
}
